shared/react-native/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java

```diff
@@ -76,27 +77,66 @@ private void createDummyFile() {
+            return null;
+        }
 
         return instanceManager.getCurrentReactContext();
     }
 
-    private void handleNotificationIntent(Intent intent) {
-        if (!intent.getBooleanExtra("isNotification", false)) return;
-        intent.removeExtra("isNotification");
-
-        ReactContext reactContext = getReactContext();
-        if (reactContext == null) return;
-
+    private void handleNotificationIntentWithContext(Intent intent, ReactContext reactContext) {
         DeviceEventManagerModule.RCTDeviceEventEmitter emitter = reactContext
                 .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class);
-        if (emitter != null) {
-            WritableMap evt = Arguments.createMap();
-            evt.putString("convID", intent.getStringExtra("convID"));
-            evt.putString("type", "chat.newmessage");
-            evt.putBoolean("userInteraction", true);
-            emitter.emit("androidIntentNotification", evt);
+        if (emitter == null) {
+            NativeLogger.warn("notification emitter not ready");
+            return;
         }
+        WritableMap evt = Arguments.createMap();
+        evt.putString("convID", intent.getStringExtra("convID"));
+        evt.putString("type", "chat.newmessage");
+        evt.putBoolean("userInteraction", true);
+        emitter.emit("androidIntentNotification", evt);
+    }
+
+    private void handleNotificationIntent(final Intent intent) {
+        if (!intent.getBooleanExtra("isNotification", false)) return;
+        intent.removeExtra("isNotification");
+        NativeLogger.info("launching from notification");
+
+        final ReactContext reactContext = getReactContext();
+        if (reactContext == null) {
+            NativeLogger.warn("react context not ready");
+            (new Thread(new Runnable() {
+                @Override
+                public void run() {
+
+                    // Try for 10 seconds
+                    int millisecondsToSleep = 18000;
+                    int millisecondsPerSleep = 6000;
+                    int numSleeps = millisecondsToSleep / millisecondsPerSleep;
+                    for (int i=0; i < numSleeps; i++) {
+                        try {
+                            Thread.sleep(millisecondsPerSleep);
+                        } catch (InterruptedException ex) {
+                            Thread.currentThread().interrupt();
+                        }
+                        final ReactContext reactContext = getReactContext();
+                        if (reactContext == null) {
+                            continue;
+                        }
+                        // FIXME: Currently this is very sloppy. Even after reactContext is no
+                        // longer null, it's possible to error when trying to route from the
+                        // resulting action.
+                        handleNotificationIntentWithContext(intent, reactContext);
```

Comment:
@MarcoPolo mind taking a look here? Ideally I'd wait for the React lifecycle to be done loading and past the login screen; right now I can only play with the timing.

Reply:
I don't think the mainactivity should be doing the event emitting. That's why you're having these problems. If you move the emitter to a native module, I don't think you'll have this problem.

Try adding something like this to your push handler module:

```java
public class PushHandler extends ReactContextBaseJavaModule {
    ...
    public PushHandler(final ReactApplicationContext reactContext) {
        super(reactContext);
        // You might have to do this here in case the lifecycle event below doesn't fire on first launch.
        // Important: do not store a ref to current activity
        handleIntent(this.getCurrentActivity().getIntent()); // Move all the logic out of mainactivity into this module now


        reactContext.addLifecycleEventListener(new LifecycleEventListener() {
            @Override
            public void onHostResume() {
              // Important: do not store a ref to current activity
             handleIntent(this.getCurrentActivity().getIntent());
            }
        });
    }
}
```

I think that should work because this module will only be created when react context is ready.

Reply:
I will try that. Thanks!

Reply:
@MarcoPolo:
```
03-26 11:08:42.607 32165 32275 E unknown:ReactNative: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Intent android.app.Activity.getIntent()' on a null object reference
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at io.keybase.ossifrage.modules.PushHandler.<init>(PushHandler.java:137)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at io.keybase.ossifrage.KBReactPackage.createNativeModules(KBReactPackage.java:40)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at com.facebook.react.NativeModuleRegistryBuilder.processPackage(NativeModuleRegistryBuilder.java:85)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at com.facebook.react.ReactInstanceManager.processPackage(ReactInstanceManager.java:1173)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at com.facebook.react.ReactInstanceManager.processPackages(ReactInstanceManager.java:1143)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at com.facebook.react.ReactInstanceManager.createReactContext(ReactInstanceManager.java:1085)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at com.facebook.react.ReactInstanceManager.access$900(ReactInstanceManager.java:117)
03-26 11:08:42.607 32165 32275 E unknown:ReactNative:   at com.facebook.react.ReactInstanceManager$5.run(ReactInstanceManager.java:916)
```